### ISSUE TITLE
VIT: WARNING:absl:Found untraced functions... These functions will not be directly callable after loading.

### ISSUE THREAD

**Issue by exx8** [tensorflow/models#10562]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [X] I am using the latest TensorFlow Model Garden release and TensorFlow 2.
- [X] I am reporting the issue to the correct repository. (Model Garden official or research directory)
- [X]I checked to make sure that this issue has not been filed already.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/blob/master/official/projects/vit/modeling/vit.py
## 2. Describe the bug

Once loading vit I get the following error:
```
WARNING:tensorflow:Model failed to serialize as JSON. Ignoring... 
Layer Encoder has arguments ['self', 'num_layers', 'mlp_dim', 'num_heads', 'dropout_rate', 'attention_dropout_rate', 'kernel_regularizer', 'inputs_positions', 'init_stochastic_depth_rate', 'kernel_initializer', 'add_pos_embed']
in `__init__` and therefore must override `get_config()`.

Example:

class CustomLayer(keras.layers.Layer):
    def __init__(self, arg1, arg2):
        super().__init__()
        self.arg1 = arg1
        self.arg2 = arg2

    def get_config(self):
        config = super().get_config()
        config.update({
            "arg1": self.arg1,
            "arg2": self.arg2,
        })
        return config
WARNING:tensorflow:Model failed to serialize as JSON. Ignoring... 
Layer Encoder has arguments ['self', 'num_layers', 'mlp_dim', 'num_heads', 'dropout_rate', 'attention_dropout_rate', 'kernel_regularizer', 'inputs_positions', 'init_stochastic_depth_rate', 'kernel_initializer', 'add_pos_embed']
in `__init__` and therefore must override `get_config()`.

Example:

class CustomLayer(keras.layers.Layer):
    def __init__(self, arg1, arg2):
        super().__init__()
        self.arg1 = arg1
        self.arg2 = arg2

    def get_config(self):
        config = super().get_config()
        config.update({
            "arg1": self.arg1,
            "arg2": self.arg2,
        })
        return config
2022-03-27 15:48:13.914733: W tensorflow/stream_executor/gpu/asm_compiler.cc:111] *** WARNING *** You are using ptxas 10.1.243, which is older than 11.1. ptxas before 11.1 is known to miscompile XLA code, leading to incorrect results or invalid-address errors.
```
Once I try to save the model
`model.save("/sda/Eran/imagenet-in-np/1.2M-Transformer-test")
`
I get:
`2022-03-27 15:52:29.493875: W tensorflow/python/util/util.cc:368] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING:absl:Found untraced functions such as posembed_input_layer_call_fn, posembed_input_layer_call_and_return_conditional_losses, dropout_layer_call_fn, dropout_layer_call_and_return_conditional_losses, layer_normalization_layer_call_fn while saving (showing 5 of 975). These functions will not be directly callable after loading.
`
## 3. Steps to reproduce

import vit.
get the 1st error.
train the model.
try to save the model,
get the 2nd error.

I've noticed that some of the weight remains unchanged. Maybe it is related to the errors?
## 4. Expected behavior

The model should be loaded succesfully; all the wights should have some flexibility during training
## 5. Additional context

Include any logs that would be helpful to diagnose the problem.

## 6. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): linux 20.04
- Mobile device name if the issue happens on a mobile device:X
- TensorFlow installed from (source or binary): tensorflow 2.8
- TensorFlow version (use command below):
- Python version: 3.8.10
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:
Sun Mar 27 16:01:41 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:17:00.0 Off |                  N/A |
| 27%   31C    P8    17W / 250W |      5MiB / 11264MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  On   | 00000000:65:00.0 Off |                  N/A |
| 27%   31C    P8     1W / 250W |     68MiB / 11264MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

**Comment by saberkun** [comment]
Hi, we are OK to accept your fix. However, according to keras team, missing get_config will still recover the model as a "black-box". As long as you do not inspect the inner components and use sub layers, this should be fine for finetuning.
If not, please file an issue to keras team.
Overall, as a library, when arguments are added to common layers and models, it is very hard to maintain the get_config()/from_config()

**Comment by exx8** [comment]
Well I believe these are 2 different issues.
The first one is indeed solved by the PR.
but the 2nd one still hasn't be solved.
I do the following:
```
transformer = loaded_model.layers[0]
transformer.trainable = True
```
Which is a sub-layer.
when I look at tensorboard:
![image](https://user-images.githubusercontent.com/8540180/160335807-9cacdeec-5aba-4b80-a7c1-9ad1323d2dbc.png)
I see that the kernel weight stay still, which does suggest, a bug is present.

**Comment by saberkun** [comment]
The 2nd one looks like a keras bug. Could you use gradient tape and print gradients?

**Comment by exx8** [comment]
I am not quite sure how do I tape gradients during model fit.
Do you have an example?

Thanks

**Comment by saberkun** [comment]
@exx8 you can override Model's train_step: https://github.com/keras-team/keras/blob/master/keras/engine/training.py#L861
for python you can do: model.train_step = my_train_step.
For more instructions, I would like to let the author @rchao explain.

**Comment by exx8** [comment]
it might be related to the LAMB optimizer see:
https://github.com/tensorflow/addons/issues/2687

**Comment by saberkun** [comment]
The LAMD optimizer has the gradient normalization and effective lr adjustment. I encountered one case that the small variables collapsed to zeros because of the weight-decay + LR adjustment. Could you try to turn off the weight decay?

**Comment by mhorlacher** [comment]
Any updates on that? For me, this problem always occurs when using sub-classed layers in combination with a sub-classed model. See this [gist](https://colab.research.google.com/gist/mhorlacher/e172e2dc8ec10fc790cf1c3fa1931e8a/untitled8.ipynb).

**Comment by saberkun** [comment]
@mhorlacher thank you for the code example! that reveals a keras problem. Could you share this to keras github repo?
@qlzh727 Could you ask Keras team to take a look?

**Comment by mhorlacher** [comment]
Hi @saberkun - My case is not related to VIT, so I can't comment on that. I will open a new issue in the Keras repo and link to this issue.

**Comment by laxmareddyp** [comment]
https://github.com/tensorflow/models/issues/10587
